File: agents/technical_analyst/indicators.py
"""Vectorised technical indicators implemented with pandas/NumPy.

These are pure-Python implementations so the system does not depend on the
TA-Lib C library. All functions accept a price ``Series`` (or OHLC frame) and
return pandas objects indexed identically to the input.

Conventions:
    * RSI/ATR/ADX use Wilder's smoothing.
    * Bollinger Bands use a population standard deviation (``ddof=0``).
"""

from __future__ import annotations

import numpy as np
import pandas as pd

__all__ = [
    "adx",
    "atr",
    "bollinger_bands",
    "ema",
    "macd",
    "rsi",
    "sma",
    "stochastic",
]


def sma(series: pd.Series, period: int = 20) -> pd.Series:
    """Simple moving average."""
    _validate_period(period)
    return series.rolling(window=period, min_periods=period).mean()


def ema(series: pd.Series, period: int = 20) -> pd.Series:
    """Exponential moving average (``adjust=False``)."""
    _validate_period(period)
    return series.ewm(span=period, adjust=False, min_periods=period).mean()


def rsi(series: pd.Series, period: int = 14) -> pd.Series:
    """Relative Strength Index using Wilder's smoothing."""
    _validate_period(period)
    delta = series.diff()
    gain = delta.clip(lower=0.0)
    loss = -delta.clip(upper=0.0)
    avg_gain = gain.ewm(alpha=1.0 / period, adjust=False, min_periods=period).mean()
    avg_loss = loss.ewm(alpha=1.0 / period, adjust=False, min_periods=period).mean()
    rs = avg_gain / avg_loss
    result = 100.0 - (100.0 / (1.0 + rs))
    result = result.where(avg_loss != 0.0, 100.0)
    return result.where(~((avg_loss == 0.0) & (avg_gain == 0.0)), 50.0)


def macd(
    series: pd.Series,
    fast: int = 12,
    slow: int = 26,
    signal: int = 9,
) -> pd.DataFrame:
    """Moving Average Convergence Divergence.

    Returns:
        Frame with ``macd``, ``signal`` and ``histogram`` columns.
    """
    _validate_period(fast)
    _validate_period(slow)
    _validate_period(signal)
    if fast >= slow:
        raise ValueError("fast period must be smaller than slow period")
    macd_line = ema(series, fast) - ema(series, slow)
    signal_line = macd_line.ewm(span=signal, adjust=False, min_periods=signal).mean()
    return pd.DataFrame(
        {
            "macd": macd_line,
            "signal": signal_line,
            "histogram": macd_line - signal_line,
        },
        index=series.index,
    )


def true_range(high: pd.Series, low: pd.Series, close: pd.Series) -> pd.Series:
    """True Range: max of high-low, |high-prev close|, |low-prev close|."""
    prev_close = close.shift(1)
    ranges = pd.concat(
        [high - low, (high - prev_close).abs(), (low - prev_close).abs()],
        axis=1,
    )
    return ranges.max(axis=1)


def atr(high: pd.Series, low: pd.Series, close: pd.Series, period: int = 14) -> pd.Series:
    """Average True Range using Wilder's smoothing."""
    _validate_period(period)
    tr = true_range(high, low, close)
    return tr.ewm(alpha=1.0 / period, adjust=False, min_periods=period).mean()


def bollinger_bands(series: pd.Series, period: int = 20, num_std: float = 2.0) -> pd.DataFrame:
    """Bollinger Bands.

    Returns:
        Frame with ``middle``, ``upper`` and ``lower`` columns.
    """
    _validate_period(period)
    middle = sma(series, period)
    std = series.rolling(window=period, min_periods=period).std(ddof=0)
    return pd.DataFrame(
        {
            "middle": middle,
            "upper": middle + num_std * std,
            "lower": middle - num_std * std,
        },
        index=series.index,
    )


def stochastic(
    high: pd.Series,
    low: pd.Series,
    close: pd.Series,
    k_period: int = 14,
    d_period: int = 3,
) -> pd.DataFrame:
    """Stochastic oscillator (%K / %D).

    Returns:
        Frame with ``k`` and ``d`` columns.
    """
    _validate_period(k_period)
    _validate_period(d_period)
    lowest = low.rolling(window=k_period, min_periods=k_period).min()
    highest = high.rolling(window=k_period, min_periods=k_period).max()
    span = (highest - lowest).replace(0.0, np.nan)
    k = 100.0 * (close - lowest) / span
    return pd.DataFrame({"k": k, "d": k.rolling(window=d_period, min_periods=d_period).mean()})


def adx(
    high: pd.Series,
    low: pd.Series,
    close: pd.Series,
    period: int = 14,
) -> pd.Series:
    """Average Directional Index (Wilder)."""
    _validate_period(period)
    up_move = high.diff()
    down_move = -low.diff()
    plus_dm = pd.Series(
        np.where((up_move > down_move) & (up_move > 0), up_move, 0.0), index=high.index
    )
    minus_dm = pd.Series(
        np.where((down_move > up_move) & (down_move > 0), down_move, 0.0), index=high.index
    )
    tr = true_range(high, low, close)
    atr_values = tr.ewm(alpha=1.0 / period, adjust=False, min_periods=period).mean()
    plus_di = (
        100.0
        * plus_dm.ewm(alpha=1.0 / period, adjust=False, min_periods=period).mean()
        / atr_values.replace(0.0, np.nan)
    )
    minus_di = (
        100.0
        * minus_dm.ewm(alpha=1.0 / period, adjust=False, min_periods=period).mean()
        / atr_values.replace(0.0, np.nan)
    )
    denominator = (plus_di + minus_di).replace(0.0, np.nan)
    dx = 100.0 * (plus_di - minus_di).abs() / denominator
    return dx.ewm(alpha=1.0 / period, adjust=False, min_periods=period).mean()


def _validate_period(period: int) -> None:
    if period < 2:
        raise ValueError(f"period must be at least 2, got {period}")

File: frontend/src/lib/indicators.ts
import type { OhlcBar } from "@/lib/types";

/** Client-side technical indicators used for chart overlays. */

export type Nullable = number | null;

export function sma(values: number[], period: number): Nullable[] {
  const out: Nullable[] = new Array(values.length).fill(null);
  if (period <= 0) return out;
  let sum = 0;
  for (let i = 0; i < values.length; i += 1) {
    sum += values[i];
    if (i >= period) sum -= values[i - period];
    if (i >= period - 1) out[i] = sum / period;
  }
  return out;
}

export function ema(values: number[], period: number): Nullable[] {
  const out: Nullable[] = new Array(values.length).fill(null);
  if (period <= 0 || values.length < period) return out;
  const k = 2 / (period + 1);
  let previous = 0;
  for (let i = 0; i < values.length; i += 1) {
    if (i < period - 1) {
      previous += values[i];
      continue;
    }
    if (i === period - 1) {
      previous = (previous + values[i]) / period;
      out[i] = previous;
      continue;
    }
    previous = values[i] * k + previous * (1 - k);
    out[i] = previous;
  }
  return out;
}

export function rsi(values: number[], period = 14): Nullable[] {
  const out: Nullable[] = new Array(values.length).fill(null);
  if (values.length <= period) return out;
  let gain = 0;
  let loss = 0;
  for (let i = 1; i <= period; i += 1) {
    const change = values[i] - values[i - 1];
    if (change >= 0) gain += change;
    else loss -= change;
  }
  let avgGain = gain / period;
  let avgLoss = loss / period;
  out[period] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  for (let i = period + 1; i < values.length; i += 1) {
    const change = values[i] - values[i - 1];
    const up = change > 0 ? change : 0;
    const down = change < 0 ? -change : 0;
    avgGain = (avgGain * (period - 1) + up) / period;
    avgLoss = (avgLoss * (period - 1) + down) / period;
    out[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  }
  return out;
}

export function macd(
  values: number[],
  fast = 12,
  slow = 26,
  signalPeriod = 9,
): { macd: Nullable[]; signal: Nullable[]; histogram: Nullable[] } {
  const fastEma = ema(values, fast);
  const slowEma = ema(values, slow);
  const macdLine: Nullable[] = values.map((_, i) =>
    fastEma[i] !== null && slowEma[i] !== null ? (fastEma[i] as number) - (slowEma[i] as number) : null,
  );
  const compact = macdLine.filter((value): value is number => value !== null);
  const signalCompact = ema(compact, signalPeriod);
  const signalLine: Nullable[] = new Array(values.length).fill(null);
  let cursor = 0;
  for (let i = 0; i < macdLine.length; i += 1) {
    if (macdLine[i] !== null) {
      signalLine[i] = signalCompact[cursor] ?? null;
      cursor += 1;
    }
  }
  const histogram = macdLine.map((value, i) =>
    value !== null && signalLine[i] !== null ? value - (signalLine[i] as number) : null,
  );
  return { macd: macdLine, signal: signalLine, histogram };
}

export function bollinger(
  values: number[],
  period = 20,
  mult = 2,
): { upper: Nullable[]; middle: Nullable[]; lower: Nullable[] } {
  const middle = sma(values, period);
  const upper: Nullable[] = new Array(values.length).fill(null);
  const lower: Nullable[] = new Array(values.length).fill(null);
  for (let i = period - 1; i < values.length; i += 1) {
    const window = values.slice(i - period + 1, i + 1);
    const mean = middle[i] as number;
    const variance = window.reduce((acc, value) => acc + (value - mean) ** 2, 0) / period;
    const std = Math.sqrt(variance);
    upper[i] = mean + mult * std;
    lower[i] = mean - mult * std;
  }
  return { upper, middle, lower };
}

export function atr(bars: OhlcBar[], period = 14): Nullable[] {
  const out: Nullable[] = new Array(bars.length).fill(null);
  if (bars.length <= period) return out;
  const trueRanges: number[] = bars.map((bar, i) => {
    if (i === 0) return bar.high - bar.low;
    const previousClose = bars[i - 1].close;
    return Math.max(
      bar.high - bar.low,
      Math.abs(bar.high - previousClose),
      Math.abs(bar.low - previousClose),
    );
  });
  let sum = 0;
  for (let i = 0; i < period; i += 1) sum += trueRanges[i];
  let previous = sum / period;
  out[period - 1] = previous;
  for (let i = period; i < bars.length; i += 1) {
    previous = (previous * (period - 1) + trueRanges[i]) / period;
    out[i] = previous;
  }
  return out;
}

export function stochastic(
  bars: OhlcBar[],
  kPeriod = 14,
  dPeriod = 3,
): { k: Nullable[]; d: Nullable[] } {
  const k: Nullable[] = new Array(bars.length).fill(null);
  for (let i = kPeriod - 1; i < bars.length; i += 1) {
    const window = bars.slice(i - kPeriod + 1, i + 1);
    const highest = Math.max(...window.map((bar) => bar.high));
    const lowest = Math.min(...window.map((bar) => bar.low));
    const span = highest - lowest;
    k[i] = span === 0 ? 50 : ((bars[i].close - lowest) / span) * 100;
  }
  const compact = k.filter((value): value is number => value !== null);
  const dCompact = sma(compact, dPeriod);
  const d: Nullable[] = new Array(bars.length).fill(null);
  let cursor = 0;
  for (let i = 0; i < k.length; i += 1) {
    if (k[i] !== null) {
      d[i] = dCompact[cursor] ?? null;
      cursor += 1;
    }
  }
  return { k, d };
}

/** Simple period-over-period returns. */
export function returns(values: number[]): number[] {
  const out: number[] = [];
  for (let i = 1; i < values.length; i += 1) {
    const previous = values[i - 1];
    out.push(previous === 0 ? 0 : (values[i] - previous) / previous);
  }
  return out;
}

/** Pearson correlation of two equal-length series (0 when undefined). */
export function pearson(left: number[], right: number[]): number {
  const length = Math.min(left.length, right.length);
  if (length < 2) return 0;
  const a = left.slice(-length);
  const b = right.slice(-length);
  const meanA = a.reduce((sum, value) => sum + value, 0) / length;
  const meanB = b.reduce((sum, value) => sum + value, 0) / length;
  let covariance = 0;
  let varianceA = 0;
  let varianceB = 0;
  for (let i = 0; i < length; i += 1) {
    const da = a[i] - meanA;
    const db = b[i] - meanB;
    covariance += da * db;
    varianceA += da * da;
    varianceB += db * db;
  }
  const denominator = Math.sqrt(varianceA * varianceB);
  return denominator === 0 ? 0 : covariance / denominator;
}

/** Simple moving average of a nullable series (for indicator smoothing). */
export function smaNullable(values: Nullable[], period: number): Nullable[] {
  const out: Nullable[] = new Array(values.length).fill(null);
  for (let i = period - 1; i < values.length; i += 1) {
    let sum = 0;
    let count = 0;
    for (let j = i - period + 1; j <= i; j += 1) {
      const value = values[j];
      if (value !== null) {
        sum += value;
        count += 1;
      }
    }
    if (count === period) out[i] = sum / period;
  }
  return out;
}
